### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by erikrozendaal**
Combined these changes add about 10% performance to trust anchor validation in the RPKI validator 3.

The changes mostly reflect reducing the number of allocations or unnecessary copies.

### CODE REVIEW

File: src/main/java/net/ripe/rpki/commons/crypto/x509cert/X509CertificateParser.java [hunk 1]
```diff
@@ -154,18 +153,10 @@ private static X509Certificate parseEncoded(byte[] encoded, ValidationResult res
...
-                final InputStream input = closer.register(new ByteArrayInputStream(encoded));
-                final CertificateFactory factory = CertificateFactory.getInstance("X.509");
-                return (X509Certificate) factory.generateCertificate(input);
-            } catch (final CertificateException e) {
-                return null;
-            } catch (final Throwable t) {
-                throw closer.rethrow(t);
-            }
-        } catch (final IOException e) {
+        try (InputStream input = new ByteArrayInputStream(encoded)) {
```
Comment (wibisono) [inline src/main/java/net/ripe/rpki/commons/crypto/x509cert/X509CertificateParser.java:156]: I found similar `Closer.create()` pattern on X509Crl.java, maybe we also want to do it there (for `makeX509CRLFromEncoded`)

File: src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java [hunk 2]
```diff
@@ -214,56 +196,50 @@ public boolean hasNoFailuresOrWarnings() {
...
 
     public List<ValidationCheck> getAllValidationChecksForLocation(ValidationLocation location) {
-        ArrayList<ValidationCheck> allChecks = new ArrayList<ValidationCheck>();
-        if (results.containsKey(location)) {
-            Map<ValidationStatus, List<ValidationCheck>> locationChecksMap = results.get(location);
-            allChecks.addAll(locationChecksMap.get(ValidationStatus.ERROR));
-            allChecks.addAll(locationChecksMap.get(ValidationStatus.WARNING));
-            allChecks.addAll(locationChecksMap.get(ValidationStatus.PASSED));
+        ArrayList<ValidationCheck> allChecks = new ArrayList<>();
+        ResultsPerLocation locationChecksMap = results.get(location);
```
Comment (ties) [inline src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java:238]: nit: not a `Map`

File: src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java [hunk 3]
```diff
@@ -297,17 +273,24 @@ public String toString() {
...
+            map.warning.addAll(resultsByLocation.getValue().warning);
+            map.passed.addAll(resultsByLocation.getValue().passed);
         }
         return this;
     }
+
+    private static final class ResultsPerLocation implements Serializable {
+        private static final long serialVersionUID = 1L;
+
+        final List<ValidationCheck> error = new ArrayList<>();
```
Comment (ties) [inline src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java:288]: And the JDK defaults to a zero-element/shared ArrayList, nice.

File: src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java [hunk 4]
```diff
@@ -47,12 +47,14 @@
 
 public final class ValidationResult implements Serializable {
 
-    private static final long serialVersionUID = 1L;
+    private static final long serialVersionUID = 2L;
 
-    private Map<ValidationLocation, Map<ValidationStatus, List<ValidationCheck>>> results = new TreeMap<>();
+    private Map<ValidationLocation, ResultsPerLocation> results = new TreeMap<>();
```
Comment (ties) [inline src/main/java/net/ripe/rpki/commons/validation/ValidationResult.java:52]: for reference: TreeMap is a red-black tree

File: src/main/java/net/ripe/rpki/commons/crypto/x509cert/X509CertificateParser.java (hunk 1 above)
Comment (erikrozendaal) [reply src/main/java/net/ripe/rpki/commons/crypto/x509cert/X509CertificateParser.java:156]: Fixed everywhere now.

**Pull request merged**